grails-app/services/com/unifina/service/ProductService.groovy

```diff
@@ -14,8 +15,82 @@ class ProductService {
 	ApiService apiService
 	PermissionService permissionService
 	SubscriptionService subscriptionService
+	CassandraService cassandraService
 	Random random = ThreadLocalRandom.current()
 
+	static class StreamWithMessages {
+		Stream stream
+		final List<StreamrMessage> messages = new ArrayList<>()
+		String toString() {
+			return String.format("StreamWithMessages[stream=%s, messages=%s]", stream, messages.toString())
+		}
+		StreamrMessage findMostRecentMessage() {
+			Date date = new Date(0)
+			StreamrMessage msg = null
+			for (StreamrMessage m : messages) {
+				if (m.getTimestamp().after(date)) {
+					date = m.getTimestamp()
+					msg = m
+				}
+			}
+			return msg
+		}
+	}
+
+	static class StaleProduct {
+		Product product
+		final List<StreamWithMessages> streams = new ArrayList<>()
+		String toString() {
+			return String.format("StaleProduct[product=%s, streams=%s]", product, streams.toString())
+		}
+	}
+
+	// ???This will skip duplicate product in the results in case of product with two streams???
+	boolean contains(List<StaleProduct> products, StaleProduct stale) {
+		for (StaleProduct sp : products) {
+			if (sp.product.name == stale.product.name && sp.product.id == stale.product.id) {
+				return true
+			}
+		}
+		return false
+	}
+
+	List<StaleProduct> findStaleProducts(List<Product> products, final Date threshold) {
+		// Find products and their streams
+		final List<StaleProduct> staleProducts = new ArrayList<>()
+		for (Product p : products) {
+			StaleProduct stale = new StaleProduct()
+			stale.product = p
+			// Get messages from streams
+			for (Stream s : p.getStreams()) {
+				StreamrMessage msg = cassandraService.getLatestFromAllPartitions(s)
+				if (msg != null) {
+					StreamWithMessages sm = new StreamWithMessages()
+					sm.stream = s
+					sm.messages.add(msg)
+					stale.streams.add(sm)
```

Comment:
I suggest you filter while iterating to simplify this considerably:

- Only create and add a `StreamWithMessages` to the `StaleProduct stale` if the stream is stale
- A stream if stale if the newest message is old enough + note that streams with no data should be considered stale too
- Only add `stale` to `staleProducts` if `stale.streams.size() > 0`


Something like (untested):
```
// ...
for (Stream s : p.getStreams()) {
    StreamrMessage msg = cassandraService.getLatestFromAllPartitions(s)
    if (msg != null && msg.getTimestamp().before(threshold)) { // logic corrected: after -> before!
        StreamWithLatestMessage sm = new StreamWithLatestMessage(stream: s, latestMessage:msg)
        stale.streams.add(sm)
    } else if (msg == null) {
        // Streams with no messages are considered stale, with latestMessage null
        StreamWithLatestMessage sm = new StreamWithLatestMessage(stream: s)
        stale.streams.add(sm)
    }
}

if (!stale.streams.isEmpty()) {
    staleProducts.add(stale)
}

// Remove the separate filtering phase
```